### PR TITLE
Resolve "Currency Converter API" can not get Currency Rate

### PR DESCRIPTION
**Pull request by edenduong**
### Description (*)

Resolve "Currency Converter API" can not get Currency Rate.

As my inspection, https://free.currencyconverterapi.com/ now required the API key.

image

But now in Magento 2.3, it doesn't have the API Key config. => Missing information and can not get the rate.

Solution: Add the "API Key" config to the settings, pass the key to the API Url.
- also Process the "Error" response (to fix the issue magento/magento2#24007 -**Notice: Undefined index** ) 

### Fixed Issues (if relevant)

1. magento/magento2#24007: "Currency Converter API" notify "Notice: Undefined index" when import currency rate
### Manual testing scenarios (*)

**- When no fill API Key**

1. Go to Backend
2. Store -> Currency Rate
3. "Import Service" choose "Currency Converter API"
4. Click "Import"

#### Expected result 
1. Notify correct error message 

**- When fill correct API Key**
1. Get API from https://free.currencyconverterapi.com/ (by register)
2. Store->Configurations->Currency Setup
3. Currency Converter API -> API Key. Fill the **API Key** get from email which sent by https://free.currencyconverterapi.com/
4. Store -> Currency Rate
5.  "Import Service" choose "Currency Converter API"
6. Click "Import"

#### Expected result 
1. Get the rate successfully from currencyconverterapi

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php [hunk 1]
```diff
@@ -75,10 +80,15 @@ public function fetchRates()
     private function convertBatch($data, $currencyFrom, $currenciesTo)
     {
         foreach ($currenciesTo as $to) {
+            //phpcs:ignore Magento2.Functions.DiscouragedFunction
             set_time_limit(0);
             try {
                 $url = str_replace('{{CURRENCY_FROM}}', $currencyFrom, self::CURRENCY_CONVERTER_URL);
                 $url = str_replace('{{CURRENCY_TO}}', $to, $url);
+                $url = $url . '&apiKey=' . $this->scopeConfig->getValue(
```
Comment (swnsma) [inline app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php]: Could you add 'apiKey' parameter to the constant and replace it with with the value?

File: app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php [hunk 2]
```diff
@@ -87,9 +97,14 @@ private function convertBatch($data, $currencyFrom, $currenciesTo)
                         $this->_messages[] = __('We can\'t retrieve a rate from %1 for %2.', $url, $to);
                         $data[$currencyFrom][$to] = null;
                     } else {
-                        $data[$currencyFrom][$to] = $this->_numberFormat(
-                            (double)$response[$currencyFrom . '_' . $to]
-                        );
+                        if (isset($response['error']) && $response['error']) {
```
Comment (swnsma) [inline app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php]: It is really nice to have error handling!

Can we check if API key is set in the beginning of the method?
With this we will be able to show error message about missed API key and avoid calls to API in the loop with missed API key.

**Review by swnsma — CHANGES_REQUESTED** [review]

File: app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php (hunk 2 above)
Comment (edenduong) [reply app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php]: I added it. Please check it again. Thanks.

File: app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php (hunk 1 above)
Comment (edenduong) [reply app/code/Magento/Directory/Model/Currency/Import/CurrencyConverterApi.php]: I added it. Please check it again. Thanks.

**Review by swnsma — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @swnsma, thank you for the review.
[ENGCOM-5574](https://jira.corp.magento.com/browse/ENGCOM-5574) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Hi @edenduong !

During testing we faced an issue.

**Problem:** Duplicate "_Invalid Free API Key..._" error message

**Steps to reproduce:**

1. Go to Admin->Stores->Configuration->Currency Setup;
2. Select 3 or more currencies in "Allowed Currencies" block;
3. In "Currency Converter API" block set wrong API code -> Save;
4. Go to System->Currency Rates;
5. Select "Currency Converter API" from "Import Service" dropdown;
6. Click on "Import" button;

**Actual Result:**
screen

@edenduong Could you take a look?

Thanks!

**Comment by edenduong** [comment]
Hello @engcom-Alfa,

I have changed the source code. You can check it again.

Hi @swnsma : Please review my new changes. Thanks!

**Review by swnsma — APPROVED** [review]
It seems OK for me.

**Comment by magento-engcom-team** [comment]
Hi @swnsma, thank you for the review.
[ENGCOM-5574](https://jira.corp.magento.com/browse/ENGCOM-5574) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Pull request merged**